src/hotspot/cpu/x86/peephole_x86_64.cpp

```diff
@@ -235,6 +236,132 @@ bool Peephole::test_may_remove(Block* block, int block_index, PhaseCFG* cfg_, Ph
   return false;
 }
 
+// This function removes redundant lea instructions that result from chained dereferences that
+// match to leaPCompressedOopOffset, leaP8Narow, or leaP32Narrow. This happens for ideal graphs
+// of the form LoadN -> DecodeN -> AddP. Matching with any leaP* rule consumes both the AddP and
+// the DecodeN. However, after matching the DecodeN is added back as the base for the leaP*,
+// which is nessecary if the oop derived by the leaP* gets added to an OopMap, because OopMaps
+// cannot contain derived oops with narrow oops as a base.
```

Comment:
Am I correct to assume that if it is referenced in OopMap (which is side table) it will by referenced by some Safepoint node in graph?

Reply:
Exactly. This is why I can get away with only checking the usages of the decode.

Reply:
This scenario looks like this: 
![image](https://github.com/user-attachments/assets/0c012810-c158-46dc-997d-6b5040c16377)

Test coming soon.

Reply:
I spoke too soon. I am not able to reduce a test that reliable safepoints right after a lea.